state-sync/aptos-data-client/src/client.rs

```diff
@@ -183,60 +231,218 @@ impl AptosDataClient {
     /// latency from the given set of serviceable peers.
-    fn choose_random_peer_by_distance_and_latency(
+    fn choose_random_peers_by_distance_and_latency(
         &self,
-        request: &StorageServiceRequest,
         serviceable_peers: HashSet<PeerNetworkId>,
-    ) -> Result<PeerNetworkId, Error> {
-        // Choose a peer weighted by distance and latency
-        if let Some(peer) = utils::choose_random_peer_by_distance_and_latency(
+        num_peers_to_choose: usize,
+    ) -> HashSet<PeerNetworkId> {
+        // Choose peers weighted by distance and latency
+        let selected_peers = utils::choose_random_peers_by_distance_and_latency(
             serviceable_peers.clone(),
             self.get_peers_and_metadata(),
-        ) {
-            return Ok(peer); // Return the peer if we found one
-        }
+            num_peers_to_choose,
+        );
 
-        // Otherwise, simply select a peer at random
-        self.choose_random_peer(request, serviceable_peers)
+        // Extend the selected peers with random peers (if necessary)
+        utils::extend_with_random_peers(selected_peers, serviceable_peers, num_peers_to_choose)
     }
 
-    /// Choose a connected peer that can service the given request.
-    /// Returns an error if no such peer can be found.
-    pub(crate) fn choose_peer_for_request(
+    /// Chooses several connected peers to service the given request.
+    /// Returns an error if no single peer can service the request.
+    pub(crate) fn choose_peers_for_request(
         &self,
         request: &StorageServiceRequest,
-    ) -> crate::error::Result<PeerNetworkId, Error> {
-        // All requests should be sent to prioritized peers (if possible).
-        // If none can handle the request, fall back to the regular peers.
-        let (priority_peers, regular_peers) = self.get_priority_and_regular_peers()?;
-        let priority_serviceable = self.identify_serviceable(priority_peers, request);
-        let serviceable_peers = if !priority_serviceable.is_empty() {
-            priority_serviceable
+    ) -> crate::error::Result<HashSet<PeerNetworkId>, Error> {
+        // Get all peers grouped by priorities
+        let peers_by_priorities = self.get_peers_by_priorities()?;
+
+        // Identify the peers that can service the request (ordered by priority)
+        let mut serviceable_peers_by_priorities = vec![];
+        for priority in PeerPriority::get_all_ordered_priorities() {
+            // Identify the serviceable peers for the priority
+            let peers = self.identify_serviceable(&peers_by_priorities, priority, request);
+
+            // Add the serviceable peers to the ordered list
+            serviceable_peers_by_priorities.push(peers);
+        }
+
+        // If the request is a subscription request, select a single
+        // peer (as we can only subscribe to a single peer at a time).
+        if request.data_request.is_subscription_request() {
```

Comment:
(Not really about this line in particular)

Any thoughts on reducing the `max_subscription_lag_secs` from the default 30s after this change? If multi-fetch can improve things, maybe it's better to unsubscribe more aggresively?

Reply:
Yeah, makes sense! 😄 Note: Subscription syncing is disabled, so when we enable it I'll reduce the timeout.